### PR TITLE
[PyTorch] Refactor FP8 workspaces in linear modules

### PR DESCRIPTION
**Pull request by timmoon10**
This PR refactors the logic for FP8 weight workspaces in `te.Linear`, `te.LayerNormLinear`, and `te.LayerNormMLP`. The existing logic is somewhat convoluted since it was designed to pass around raw UINT8 buffers and `Float8Tensor` support was kludged in https://github.com/NVIDIA/TransformerEngine/pull/452. For example, when `te.Linear` has FP32 params, it maintains two workspace `Float8Tensor`s (FP8 weight and transpose) and in each forward pass it will extract out the buffers to create another temporary `Float8Tensor`. This PR streamlines the process so it will just maintain a single workspace `Float8Tensor`.

Motivations:
- This fixes an FP8 recipe bug introduced in https://github.com/NVIDIA/TransformerEngine/pull/575 (see https://github.com/NVIDIA/TransformerEngine/pull/786#pullrequestreview-2024087226). The FP8 scale update kernel updates the scales for weights in every forward pass, even ones where the FP8 weights are not updated, so we can run into situations where the FP8 scales don't match the FP8 weights. This PR fixes this by taking advantage of the fact that `Float8Tensor` has a private copy of the FP8 scale-inverse won't be affected by scale updates until its values are updated.
- FP8 compute can sometimes result in performance degradation due to CPU overheads (see https://github.com/NVIDIA/TransformerEngine/issues/761). The `Float8Tensor` constructor requires a CUDA kernel launch to initialize the FP8 scale-inverse, so creating unnecessary `Float8Tensor`s adds non-trivial CPU overhead. Benchmarking the forward pass of small `te.Linear`s, this PR gives a 1.12x speedup. 
- I find the logic in this PR a bit easier to reason about, although I'd appreciate feedback. It feels nicer to let `Float8Tensor` internally handle things like FP8 casting and interacting with `fp8_meta`.

### CODE REVIEW

**Comment by timmoon10** [comment]
This PR has grown in scope as I've identified bugs:
- Our current API for transpose caching in `Float8Tensor` doesn't work well with CUDA graphs. The only way to access the cache is to call `transpose_2d(cache=True)`, which fills the cache. But if you capture a CUDA graph with `is_first_microbatch=None`, this means that the cache is filled during the warmup steps and you never actually capture the transpose kernel. The easiest fix is to modify `Float8Tensor` to support lazy transpose caching (see https://github.com/NVIDIA/TransformerEngine/pull/575#discussion_r1548888680, https://github.com/NVIDIA/TransformerEngine/pull/575#pullrequestreview-1985448624, and https://github.com/NVIDIA/TransformerEngine/pull/735).
- The ONNX export tests were failing because they assume the FP8 scales can be represented with constant operations, which requires that the scales are initialized during the ONNX export. This doesn't support the case where we do training before export.

File: transformer_engine/pytorch/float8_tensor.py [hunk 1]
```diff
@@ -381,7 +382,8 @@ def __new__(
             raise ValueError(
                 "Float8Tensor requires non-differentiable data buffer"
             )
-        data = data.cuda()
+        if not data.is_cuda:
+            data = data.cuda()
```
Comment (ksivaman) [inline transformer_engine/pytorch/float8_tensor.py:386]: Is this not a noop if tensor is already on device?

Comment (timmoon10) [reply transformer_engine/pytorch/float8_tensor.py:386]: Yep, tensor conversion functions (`to`, `cuda`, `contiguous`, `float`) will be no-ops if possible. I prefer being as paranoid as possible until CPU overheads start becoming a problem.

**Comment by ksivaman** [comment]
For testing CUDA graphs with FP8 caching, did you use the `noop_flag` in `transpose` and the `fp8_weight_caching` flag in `make_graphed_callables`?

File: transformer_engine/pytorch/float8_tensor.py (hunk 1 above)
Comment (timmoon10) [reply transformer_engine/pytorch/float8_tensor.py:386]: Ah, I missed the "not" in your question. The conversion functions add some CPU overhead, so this is a microoptimization.

**Comment by timmoon10** [comment]
With the tests and bugfixes in https://github.com/NVIDIA/TransformerEngine/pull/869, this PR seems to handle `make_graphed_callables` with `fp8_weight_caching=True` correctly.

**Review by ksivaman — APPROVED** [review]
Approving given numerics for all cases remain identical in e2e tests

**Comment by timmoon10** [comment]
Training GPT for 100 steps (175B params, TP=2, PP=4), I don't see significant differences in the loss curves with and without this PR. I think this is ready to go.
